## Running node server for pendlaren
I use the package [forever](https://www.npmjs.com/package/forever) to run the node server in the background.
Running node without forever.
```
node pendlaren_node_server.js
```
stop with ctrl-c ..

Command line stuff.
This command:
forever start -l forever.log -o out.log -e err.log pendlaren_node_server.js

--
Forever is a very good NodeJs module to do exactly that.

Install forever by typing in the command line

$ npm install forever -g
Then use the following command to run a node.js script

$ forever start /path/to/script.js
You are good to go. Additionally you can run

$ forever list
to see all the running scripts. You can terminate any specific script by typing

$ forever stop [pid]
where [pid] is the process ID of the script you will obtain from the list command. To stop all scripts, you may type

$ forever stopall

#### Here are some code to start stop and list the running forever instances.

In [ ]:
#File list in current dir
! ls -lah

In [ ]:
#Run an instance with forever
! rm out.log
!  forever start -a -l forever.log -o out.log -e err.log pendlaren_node_server.js

In [ ]:
#List running forever instances
! forever list

In [ ]:
#Stop running forever instances
! forever stopall

In [ ]:
#Current run
f = open("out.log") 
print f.read()
#Older logs
# f = open("/home/lars_rauer/.forever/forever.log") 
# print f.read()

### Node commands 
Start test server

In [15]:
! node pendlaren_node_server.js

results: "Node server starting"
Model exists for: tnK534JMwwfhvUEycn69HPbhqkt2
Predicted for :tnK534JMwwfhvUEycn69HPbhqkt2 from: 81802 to: 80014 with probability: 0.99 prediction took: 0s
Model exists for: tnK534JMwwfhvUEycn69HPbhqkt2
Predicted for :tnK534JMwwfhvUEycn69HPbhqkt2 from: 81802 to: 80014 with probability: 0.99 prediction took: 0s
Model exists for: tnK534JMwwfhvUEycn69HPbhqkt2
Predicted for :tnK534JMwwfhvUEycn69HPbhqkt2 from: 81802 to: 80014 with probability: 0.99 prediction took: 0s
^C


In [ ]:
# logg all consoles to files
! node pendlaren_node_server.js > test.log 2>&1

In [2]:
# ! ps -A

In [1]:
#Python version
! python --version

Python 3.7.1


In [ ]:
! killall -9 node

### Prediction time test
Time from a node js script untill response from flask hello.py<br>
3 seconds first time<br>
235ms<br>
240ms<br>
247ms<br>
236ms<br>
232ms<br>
237ms<br>
### Retraining time test
tnk: 2.81s 160 rows<br>
tnk: 2.71s<br>
tnk: 2.84s<br>
eha: 5.44s 3470 rows<br>
eha: 5.25<br>
eha: 5.62
#### Retraing first time newly started server:
eha: 77s
Then
eha: 5.61s
### Small error testing:
#### Prediction working
Prediction done for: tnK534JMwwfhvUEycn69HPbhqkt2
Accuracy: 1.0
fromStation: 80014
toStation: 80002
#### Prediction no existing model.
Prediction error or no trained mode found for: ehaBtfOPDNZjzy1MEvjQmGo4Zv1
#### Prediction no running flask server
Error in connection for: ehaBtfOPDNZjzy1MEvjQmGo4Zv1
#### Training working
Done training for: ehaBtfOPDNZjzy1MEvjQmGo4Zv12
#### Training no id
Error in connection for: undefined
#### Training no runnig flask server
Error in connection for: ehaBtfOPDNZjzy1MEvjQmGo4Zv12
#### Training no trained model
Unknown training error for: ehaBtfOPDNZjzy1MEvjQmGo4Zv1
#### Running multiple 3 trainings same time (A user cannot train a same model when one training is on the way).
Hangs it seems with concurrent.....2 working fine....restar efter 5 min..<br>
Stopped flask server no result<br>
restart node 8min 49s<br>
Node server hangs....<br>
restart flask.. Address already in use<br>
kill flask not working still there....<br>
restart python<br>
netstat -tulnp | grep :5000<br>
cannot remove listening port<br>
restart server.. no sudo password....<br>
This could perhaps be a problem when starting server and many waiting for training<br>
#### Running multiple 3 trainings different users same time (again)
Nope<br>
#### Running multiple 3 trainings different users about same time (after)
Nope but restarting node server is enough.
#### Running multiple 2 trainings same time
works well
#### Running multiple 3 predictions same time
Works well
#### Simulating many predictions
setInterval(pred1, 2000);<br>
setInterval(pred1, 3000);<br>
setInterval(pred1, 4000);<br>
OK works well.
#### simulating many predictions 500ms between each for same model (no queing)
setInterval(pred1, 600);<br>
setInterval(pred1, 550);<br>
setInterval(pred1, 500);
#### simulating many predictions 200ms between each for same model (queing)
setInterval(pred1, 200);<br>
setInterval(pred1, 230);<br>
setInterval(pred1, 210);
Works for<br>
CPU times: user 1.59 s, sys: 208 ms, total: 1.8 s<br>
Wall time: 41.3 s<br>
15 preds/s
#### dotraining get prediction same time
getPrediction('ehaBtfOPDNZjzy1MEvjQmGo4Zv12',3,1242478163,840,1) //8115681079<br>
doTraining('ehaBtfOPDNZjzy1MEvjQmGo4Zv12')
#### Prediction while training
Blocks training that is ok--<br>
Works after
#### Predicting and training
setInterval(pred1, 1000);<br>
setInterval(pred2, 2000);<br>
setInterval(pred3, 2500);<br>
setInterval(train1, 10000);<br>
setInterval(train2, 15000);<br>
setInterval(train3, 20000);<br>
CPU times: user 3.56 s, sys: 448 ms, total: 4.01 s<br>
Wall time: 2min 4s<br>
Works well but:::<br>
CPU at 60% for compute engine Google Cloud so about max<br>
n1-highmem-2 (2 vCPUs, 13 GB memory)<br>
Intel Broadwell<br>
1 x NVIDIA Tesla K80
#### Seems to hang sometimes.....
port 5000 busy python process cannot kill proccess need sudo psw...

In [14]:
%%time
! node getWallTime

{"error": 0}
Done training for: tnK534JMwwfhvUEycn69HPbhqkt2
CPU times: user 56 ms, sys: 16 ms, total: 72 ms
Wall time: 2.57 s
